docs/specs/code/REFACTOR_PROPOSAL.md

```diff
@@ -0,0 +1,30 @@
+# Idea
+
+Refactor the app, use newer dependencies, improve code quality
+
+# Motivation
+
+Currently there are some things in the project which make working with it a bit too hard. The project structure could be better, the routing could be better, we could use the latest react-router and react and that could help a lot.
+
+
+# Proposal
+
+- update dependencies: react-router 4 instead of 3, React 16 instead of 15
+
+Basically the routing in react-router 4 is more handy and allows to work with it with less effort. As for React 16, there are some useful things - for example, `<Fragment />` component, it can be used to avoid wrapping multiple JSX elements in the unnecessary `<div />`. Also there are new lifecycle methods in React 16, and the React team works on the optimisation, so this can be useful.
+
+- update routing
+
+Currently the routing breaks DRY principle. There are just a lot of repeated things in each route. This appearance makes code hard to read and understand. Also if we need to add a new route, we have to repeat all these previously repeated things for a new route too.
+
+- use Mobx Provider in a proper way
```

Comment:
The reason we got rid of the Provider is actually because it doesn't work with `Flow` and so using a Provider causes a bunch of types to fail to resolve which has lead to sneaky bugs in the past

Reply:
I'll check out what we can do with Flow and Provider. By getting rid of the Provider we also got rid of access to the stores for the deeply nested components.